Parse/src/main/java/com/parse/ParseQuery.java

```diff
@@ -960,33 +960,21 @@ public ParseQuery(ParseQuery<T> query) {
     return ParseUser.getCurrentUserAsync();
   }
 
-  private void checkIfRunning() {
-    checkIfRunning(false);
-  }
-
-  private void checkIfRunning(boolean grabLock) {
-    synchronized (lock) {
-      if (isRunning) {
-        throw new RuntimeException(
-            "This query has an outstanding network connection. You have to wait until it's done.");
-      } else if (grabLock) {
-        isRunning = true;
-        cts = Task.create();
-      }
-    }
-  }
-
   /**
    * Cancels the current network request (if one is running).
    */
   //TODO (grantland): Deprecate and replace with CancellationTokens
```

Comment:
I think it should be implemented internally first, and keep the developer's original behavior. Replace the `Task<Void> cancellationToken` in `ParseQueryController ` with  `CancellationToken cancellationToken`.

https://github.com/parse-community/Parse-SDK-Android/blob/master/Parse/src/main/java/com/parse/ParseQueryController.java

```java
interface ParseQueryController {
  <T extends ParseObject> Task<List<T>> findAsync(ParseQuery.State<T> state, ParseUser user,
      CancellationToken cancellationToken);
  <T extends ParseObject> Task<Integer> countAsync(ParseQuery.State<T> state, ParseUser user,
      CancellationToken cancellationToken);
  <T extends ParseObject> Task<T> getFirstAsync(ParseQuery.State<T> state, ParseUser user,
      CancellationToken cancellationToken);
}
```

Then re-implement all the relevant files and pass cancellationToken to Bolt Task methods (ex. [onSuccessTask](https://github.com/BoltsFramework/Bolts-Android/blob/master/bolts-tasks/src/main/java/bolts/Task.java#L837-L840)).